### PR TITLE
support under-/overflows in YODA2 writer

### PR DESCRIPTION
**Pull request by 20DM**
Add support for under/overflow bins in YODA writer + unit test.

Skips points with infinity edges in YODA1 writer.

Needed for #55

### CODE REVIEW

**Comment by 20DM** [comment]
Hi Graeme,

apologies for the delay, I had a few more urgent deadlines squeezing in, but I've finally managed to take a look at this.

This changes set ensures that under/overflow bins are supported with the default YODA writer.

For the YODA1 writer this is less trivial: In Rivet we'd work out the binning from the reference-data objects, so adding two more "dummy points" to the Scatters will likely cause mayhem. In the YODA1 series we would also typically drop the under/overflow bins from the MC histograms when converting them to Scatters, so not adding the overflows to the reference data seems more consistent with that, even though we'd be removing "good data points". This is what I've done in this PR now. Hopefully an additional incentive to upgrade to the Rivet4/YODA series where this is properly supported.

File: hepdata_converter/testsuite/test_yoda1writer.py [hunk 1]
```diff
@@ -32,3 +32,17 @@ def test_parse_all(self, test_submissions):
                                       options={'output_format': 'yoda1'})
 
             self.assertNotEqual(os.stat(output_file_path).st_size, 0, 'output yoda file is empty')
+
+
+    @insert_path('yaml_inf')
+    @insert_data_as_file('yoda1/with_overflows.yoda')
+    def test_parse_no_dependent(self, yaml_simple_path, yoda_template):
```
Comment (GraemeWatt) [inline hepdata_converter/testsuite/test_yoda1writer.py:39]: Would this be better named something like `test_parse_with_overflows`?

**Review by GraemeWatt — CHANGES_REQUESTED** [review]
Thanks, this is great!  After I merge this PR, I'll look into making similar changes for the CSV and ROOT writers, but I've also got other things to work on first, so it might take a while.

I made some separate comments about the naming of the new test functions and that it would be good to reduce the size of the test YAML data file.  Looking at the YODA2 output, I don't see any explicit labels for the underflow/overflow bins, so how does the YODA code identify them?  Does it just count the number of edges (n_e) and values (n_v) and assume that the first and last bins are underflow and overflow if n_v = n_e + 1?  Does this mean that the case of an overflow bin without an underflow bin (or vice versa) is not supported (i.e. n_v = n_e)?  If so, I think the YODA2 code needs to handle  this case in the conversion from YAML, perhaps by just omitting the single underflow/overflow bin as in the YODA1 case.

**Comment by 20DM** [comment]
> Looking at the YODA2 output, I don't see any explicit labels for the underflow/overflow bins, so how does the YODA code identify them? Does it just count the number of edges (n_e) and values (n_v) and assume that the first and last bins are underflow and overflow if n_v = n_e + 1? Does this mean that the case of an overflow bin without an underflow bin (or vice versa) is not supported (i.e. n_v = n_e)? If so, I think the YODA2 code needs to handle this case in the conversion from YAML, perhaps by just omitting the single underflow/overflow bin as in the YODA1 case or maybe by writing a value of zero for the missing underflow/overflow bin.

Hi Graeme, continuous axes in YODA2 will always range from -inf to +inf and the list of edges that define the visible range will then partition the infinitely long axis. As a result a continuous axis will always have an under- and overflow bins, irrespective of whether these are needed in the end, so the cases you're concerned about are implicitly supported.

File: hepdata_converter/testsuite/test_yodawriter.py [hunk 2]
```diff
@@ -65,7 +65,7 @@ def test_parse_pattern_unmatch(self, yaml_simple_path, yoda_template):
 
     @insert_path('yaml_no_independent')
     @insert_data_as_file('yoda/no_independent.yoda')
-    def test_parse_no_dependent(self, yaml_simple_path, yoda_template):
+    def test_parse_with_overflows(self, yaml_simple_path, yoda_template):
```
Comment (GraemeWatt) [inline hepdata_converter/testsuite/test_yodawriter.py:68]: This function should be named `test_parse_no_independent` not `test_parse_with_overflows`.

File: hepdata_converter/testsuite/test_yodawriter.py [hunk 3]
```diff
@@ -76,3 +76,16 @@ def test_parse_no_dependent(self, yaml_simple_path, yoda_template):
         with open(output_file_path, 'r') as f:
             self.assertMultiLineAlmostEqual(f, yoda_template)
 
+    @insert_path('yaml_inf')
+    @insert_data_as_file('yoda/with_overflows.yoda')
+    def test_parse_no_dependent(self, yaml_simple_path, yoda_template):
```
[the remark this answers is not in the document]
Comment (GraemeWatt) [reply hepdata_converter/testsuite/test_yodawriter.py:81]: This function should be named `test_parse_with_overflows` not `test_parse_no_dependent`.

**Review by GraemeWatt — CHANGES_REQUESTED** [review]
Sorry to nitpick, but there is still an issue with the test function naming (see separate comments).

Also, I don't think you understood my [previous comment](https://github.com/HEPData/hepdata-converter/pull/56#pullrequestreview-1925330077), so let me be more explicit with a minimal example.  Consider two YAML data files with only two bins:

1. An underflow bin and a finite bin (0-100):
```yaml
independent_variables:
- header:
    name: x
  values:
  - low: -.inf
    high: 0.0
  - low: 0.0
    high: 100.0
dependent_variables:
- header:
    name: data
  values:
  - value: 0.1
  - value: 0.2
```
2. A finite bin (0-100) and an overflow bin.
```yaml
independent_variables:
- header:
    name: x
  values:
  - low: 0.0
    high: 100.0
  - low: 100.0
    high: .inf
dependent_variables:
- header:
    name: data
  values:
  - value: 0.1
  - value: 0.2
```
I checked that both these examples give the same YODA output with the current code.  I think it is a legitimate use case to give an overflow bin without an underflow bin (or vice versa) in the YAML format, so the YODA output would need to write a zero `value` (with no `errors`) for the missing underflow/overflow bin to remove the ambiguity.

**Comment by 20DM** [comment]
Hi Graeme,

I'm afraid I'm still a bit confused as to what the issue actually is, but it's probably just me. :-)

Let me try elaborate on my original response: All continuous axes in YODA will always come with an underflow and overflow whether they're needed or not. For the reference data we use `BinnedEstimtates`, which are initialised with a central value `nan`. The two edge cases that you point out will then look like this (for example):

```BEGIN YODA_ESTIMATE1D_V3 /REF/TEST/d01-x01-y01
IsRef: 1
Path: /REF/TEST/d01-x01-y01
Title: "Figure-aux_001"
Type: Estimate1D
---
Edges(A1): [1.200000e+03, 1.220000e+03]
ErrorLabels: ["statistical"]
# value      	errDn(1)     	errUp(1)
6.790900e-01 	-2.116800e-04	2.117400e-04
6.402400e-02 	-6.499500e-05	6.506200e-05
nan          	---          	---
END YODA_ESTIMATE1D_V3


BEGIN YODA_ESTIMATE1D_V3 /REF/TEST/d02-x01-y01
IsRef: 1
Path: /REF/TEST/d02-x01-y01
Title: "Figure-aux_002-a"
Type: Estimate1D
---
Edges(A1): [2.080000e+03, 2.100000e+03]
ErrorLabels: ["statistical"]
# value      	errDn(1)     	errUp(1)
nan          	---          	---
6.719900e-06 	-5.133800e-07	5.541800e-07
1.222500e-04 	-2.191800e-06	2.231400e-06
END YODA_ESTIMATE1D_V3
```

Here, there first one corresponds to an object with underflow but no overflow (and hence  the `nan` line comes last, i.e. the "overflow" position) and the second one corresponds to an object with overflow but no underflow (and hence  the `nan` line comes first, i.e. the "underflow" position). The two cases that you suggested do not result in the same output with the current code for me. 🤷

File: hepdata_converter/testsuite/test_yodawriter.py (hunk 2 above)
Comment (GraemeWatt) [reply hepdata_converter/testsuite/test_yodawriter.py]: You renamed the function `test_parse_no_dependent` instead of `test_parse_no_independent`, so I just pushed a fix.

**Review by GraemeWatt — APPROVED** [review]
Apologies, I must have made a mistake when checking the YODA output yesterday.  Checking again more carefully, the two YAML examples from my [previous message](https://github.com/HEPData/hepdata-converter/pull/56#pullrequestreview-1927672338) give the following YAML output:

1. An underflow bin and a finite bin (0-100):
```
BEGIN YODA_ESTIMATE1D_V3 /REF/RIVET_ANALYSIS_NAME/d01-x01-y01
IsRef: 1
Path: /REF/RIVET_ANALYSIS_NAME/d01-x01-y01
Title: "Figure-aux_001"
Type: Estimate1D
---
Edges(A1): [0.000000e+00, 1.000000e+02]
# value         
1.000000e-01    
2.000000e-01    
nan             
END YODA_ESTIMATE1D_V3
```
2. A finite bin (0-100) and an overflow bin.
```
BEGIN YODA_ESTIMATE1D_V3 /REF/RIVET_ANALYSIS_NAME/d01-x01-y01
IsRef: 1
Path: /REF/RIVET_ANALYSIS_NAME/d01-x01-y01
Title: "Figure-aux_001"
Type: Estimate1D
---
Edges(A1): [0.000000e+00, 1.000000e+02]
# value         
nan             
1.000000e-01    
2.000000e-01    
END YODA_ESTIMATE1D_V3
```
So the `nan` value is in a different position for the two cases as you kindly explained.  Sorry, the misunderstanding was on my side!

I'll merge now.  Many thanks again for your contribution!

**Comment by 20DM** [comment]
Hi Graeme - we were just trying to use this feature for an ATLAS measurement but it looks like this never actually got deployed, right?

**Comment by GraemeWatt** [comment]
> Hi Graeme - we were just trying to use this feature for an ATLAS measurement but it looks like this never actually got deployed, right?

Right, the deployed converter still corresponds to the [0.3.0 Release](https://github.com/HEPData/hepdata-converter/releases/tag/0.3.0) from 6th November 2023.  Underflow/overflow bins are still not officially supported (HEPData/hepdata#358).  I was waiting to complete the modifications for the CSV/ROOT export (#55) before making a new converter release.  The validator should also make an additional check (HEPData/hepdata-validator#56).  I'll try to come back to this soon.  In the meantime, underflow/overflow bins can be specified in the YAML upload and the visualisation will work, although I understand that you would also like the YODA export option to work properly.

**Comment by 20DM** [comment]
OK thanks. It looks like the export to YODA with v0.3.0 would crash actually, whereas it works fine with the main branch using a local installation of the `hepdata-coverter`. It would be good if we can address that in the coming weeks. I'll see if I can help with some of those open issues.

**Comment by jonbutterworth** [comment]
Hi @GraemeWatt any news on when this might be deplyed? We need it for https://arxiv.org/abs/2503.19836
Cheers, Jon
